tests/test_multipart.py

```diff
@@ -146,38 +158,125 @@ def test_multipart_encode_files_allows_filenames_as_none():
+def test_multipart_encode_files_allows_str_content() -> None:
     files = {"file": ("test.txt", "<string content>", "text/plain")}
     with mock.patch("os.urandom", return_value=os.urandom(16)):
         boundary = binascii.hexlify(os.urandom(16)).decode("ascii")
 
-        stream = encode(data={}, files=files)
+        stream = cast(MultipartStream, encode(data={}, files=files))
+        assert stream.can_replay()
 
         assert stream.content_type == f"multipart/form-data; boundary={boundary}"
-        assert stream.body == (
+        content = (
             '--{0}\r\nContent-Disposition: form-data; name="file"; '
             'filename="test.txt"\r\n'
             "Content-Type: text/plain\r\n\r\n<string content>\r\n"
             "--{0}--\r\n"
             "".format(boundary).encode("ascii")
         )
+        assert stream.get_headers()["Content-Length"] == str(len(content))
+        assert b"".join(stream) == content
+
+
+def test_multipart_encode_non_seekable_filelike() -> None:
+    """
+    Test that special readable but non-seekable filelike objects are supported,
+    at the cost of reading them into memory at most once.
+    """
+
+    class IteratorIO(io.IOBase):
+        def __init__(self, iterator: Iterator[bytes]) -> None:
+            self._iterator = iterator
+
+        def seekable(self) -> bool:
+            return False
+
+        def read(self, *args: Any) -> bytes:
+            return b"".join(self._iterator)
+
+    def data() -> Iterator[bytes]:
+        yield b"Hello"
+        yield b"World"
+
+    fileobj = IteratorIO(data())
+    files = {"file": fileobj}
+    stream = encode(files=files, boundary=b"+++")
+    assert not stream.can_replay()
+
+    content = (
+        b"--+++\r\n"
+        b'Content-Disposition: form-data; name="file"; filename="upload"\r\n'
+        b"Content-Type: application/octet-stream\r\n"
+        b"\r\n"
+        b"HelloWorld\r\n"
+        b"--+++--\r\n"
+    )
+    assert stream.get_headers() == {
+        "Content-Type": "multipart/form-data; boundary=+++",
+        "Content-Length": str(len(content)),
+    }
+    assert b"".join(stream) == content
+
+
+def test_multipart_file_streaming_memory(tmp_path: Any) -> None:
+    """
+    Test that multipart file uploads are effectively streaming, i.e. they don't
+    result in loading the entire file into memory.
+    """
+    path = str(tmp_path / "name.txt")
+
+    # Flush a relatively large file to disk to read from.
+    ONE_MB = 1024 * 1024
+    size_mb = 1
+    with open(path, "wb") as out:
+        out.write(os.urandom(int(size_mb * ONE_MB)))
+
+    def bench() -> None:
+        files = {"file": open(path, "rb")}
+        stream = encode(files=files, boundary=b"+++")
+        # Consume the stream one chunk at a time.
+        for _ in stream:
+            pass
+
+    # Measure memory usage of `main()` -- one entry per LOC (plus init/teardown).
+    memory_per_line: List[float] = memory_profiler.memory_usage((bench, (), {}))
```

Comment:
I'd suggest that we might not want to enforce memory profiling in the test case, as it might be a bit brittle, adds a requirement we could possibly do without.

Reply:
Yeah looking back at this PR, this test is *fancy* for sure. I still think it's nice to have some kind of proof that streaming works as expected (reduced memory consumption is the main motivation factor in the first place), but not absolute on keeping this either, since once we use generators there's not much risk we buffer everything up.